File: test/go/widget/refund_order_test.go
package widget_test

import (
	"context"
	"encoding/json"
	"fmt"
	"testing"
	"time"

	widget "github.com/dana-id/dana-go/v2/widget/v1"
	"github.com/google/uuid"

	"uat-script/helper"
)

const (
	refundOrderTitleCase      = "RefundOrder"
	refundOrderJsonPath       = "../../../resource/request/components/Widget.json"
	paymentForRefundTitleCase = "Payment"
)

// createTestWidgetPaymentForRefund creates a test widget payment that can be refunded
func createTestWidgetPaymentForRefund() (string, error) {
	var partnerReferenceNo string
	result, err := helper.RetryOnInconsistentRequest(func() (interface{}, error) {
		// Get the request data from the Payment JSON section
		caseName := "PaymentSuccess"
		jsonDict, err := helper.GetRequest(refundOrderJsonPath, paymentForRefundTitleCase, caseName)
		if err != nil {
			return "", err
		}

		// Set a unique partner reference number
		partnerReferenceNo = uuid.New().String()
		jsonDict["partnerReferenceNo"] = partnerReferenceNo
		jsonDict["validUpTo"] = helper.GenerateFormattedDate(900, 7)

		// Create the WidgetPaymentRequest object and populate it with JSON data
		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			return "", err
		}

		var request widget.WidgetPaymentRequest
		err = json.Unmarshal(jsonBytes, &request)
		if err != nil {
			return "", err
		}

		// Make the API call to create a payment
		ctx := context.Background()
		_, httpResponse, err := helper.ApiClient.WidgetAPI.WidgetPayment(ctx).WidgetPaymentRequest(request).Execute()
		if err != nil {
			return "", err
		}
		defer httpResponse.Body.Close()

		return partnerReferenceNo, nil
	}, 3, 2*time.Second)
	if err != nil {
		return "", err
	}
	return result.(string), nil
}

// RefundOrder
func TestRefundOrderValidScenario(t *testing.T) {
	helper.RetryTest(t, 3, 1, func() error {
		partnerReferenceNo, err := createTestWidgetPaymentPaid()
		if err != nil {
			return fmt.Errorf("failed to create paid widget payment: %w", err)
		}

		time.Sleep(2 * time.Second)

		caseName := "RefundOrderValidScenario"
		jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
		if err != nil {
			return fmt.Errorf("failed to get request data: %w", err)
		}

		jsonDict["originalPartnerReferenceNo"] = partnerReferenceNo
		jsonDict["partnerRefundNo"] = partnerReferenceNo
		jsonDict["merchantId"] = helper.TestConfig.MerchantID

		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			return fmt.Errorf("failed to marshal JSON: %w", err)
		}

		var request widget.RefundOrderRequest
		if err = json.Unmarshal(jsonBytes, &request); err != nil {
			return fmt.Errorf("failed to unmarshal JSON: %w", err)
		}

		ctx := context.Background()
		apiResponse, httpResponse, err := helper.ApiClient.WidgetAPI.RefundOrder(ctx).RefundOrderRequest(request).Execute()
		if err != nil {
			return fmt.Errorf("API request failed: %w", err)
		}
		defer httpResponse.Body.Close()

		responseJSON, err := apiResponse.MarshalJSON()
		if err != nil {
			return fmt.Errorf("failed to convert response to JSON: %w", err)
		}

		variableDict := map[string]interface{}{
			"partnerReferenceNo": partnerReferenceNo,
		}

		if err = helper.AssertResponse(
			refundOrderJsonPath,
			refundOrderTitleCase,
			caseName,
			string(responseJSON),
			variableDict,
		); err != nil {
			return err
		}
		return nil
	})
}
func TestRefundInProcess(t *testing.T) {
	t.Skip("Skip: Requires a paid order to refund, which needs complex setup with payment completion")
	caseName := "RefundInProcess"

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Marshal to JSON and unmarshal to widget SDK struct
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	var request widget.RefundOrderRequest
	err = json.Unmarshal(jsonBytes, &request)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute the SDK API call
	ctx := context.Background()
	apiResponse, httpResponse, err := helper.ApiClient.WidgetAPI.RefundOrder(ctx).RefundOrderRequest(request).Execute()
	if err != nil {
		t.Fatalf("API request failed: %v", err)
	}
	defer httpResponse.Body.Close()

	// Convert the response to JSON for assertion
	responseJSON, err := apiResponse.MarshalJSON()
	if err != nil {
		t.Fatalf("Failed to convert response to JSON: %v", err)
	}

	// Assert the success response
	variableDict := map[string]interface{}{
		"partnerReferenceNo": jsonDict["originalPartnerReferenceNo"],
	}

	err = helper.AssertResponse(
		refundOrderJsonPath,
		refundOrderTitleCase,
		caseName,
		string(responseJSON),
		variableDict,
	)
	if err != nil {
		t.Fatal(err)
	}
}

func TestRefundFailDuplicateRequest(t *testing.T) {
	helper.RetryTest(t, 3, 1, func() error {
		partnerReferenceNo, err := createTestWidgetPaymentPaid()
		if err != nil {
			return fmt.Errorf("failed to create paid widget payment: %w", err)
		}

		time.Sleep(2 * time.Second)

		caseName := "RefundFailDuplicateRequest"
		jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
		if err != nil {
			return fmt.Errorf("failed to get request data: %w", err)
		}

		jsonDict["originalPartnerReferenceNo"] = partnerReferenceNo
		jsonDict["partnerRefundNo"] = partnerReferenceNo
		jsonDict["merchantId"] = helper.TestConfig.MerchantID

		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			return fmt.Errorf("failed to marshal JSON: %w", err)
		}

		var request widget.RefundOrderRequest
		if err = json.Unmarshal(jsonBytes, &request); err != nil {
			return fmt.Errorf("failed to unmarshal JSON: %w", err)
		}

		ctx := context.Background()

		// First refund with this partnerRefundNo — should succeed.
		_, firstHTTP, err := helper.ApiClient.WidgetAPI.RefundOrder(ctx).RefundOrderRequest(request).Execute()
		if err != nil {
			return fmt.Errorf("first refund should succeed before duplicate attempt: %w", err)
		}
		firstHTTP.Body.Close()

		time.Sleep(2 * time.Second)

		// Same partnerRefundNo, different refundAmount → Inconsistent Request (4045818).
		request.SetRefundAmount(widget.Money{Value: "2.00", Currency: "IDR"})

		_, httpResponse, err := helper.ApiClient.WidgetAPI.RefundOrder(ctx).RefundOrderRequest(request).Execute()
		if err != nil {
			variableDict := map[string]interface{}{
				"partnerReferenceNo": partnerReferenceNo,
			}
			if err = helper.AssertFailResponse(refundOrderJsonPath, refundOrderTitleCase, caseName, httpResponse, variableDict); err != nil {
				return err
			}
			return nil
		}

		httpResponse.Body.Close()
		return fmt.Errorf("expected duplicate refund error for case %s but API call succeeded", caseName)
	})
}
func TestRefundFailOrderNotPaid(t *testing.T) {
	caseName := "RefundFailOrderNotPaid"

	// Create a test payment first (which will be in INIT status, not paid)
	partnerReferenceNo, err := createTestWidgetPaymentForRefund()
	if err != nil {
		t.Fatalf("Failed to create test widget payment: %v", err)
	}

	// Give time for the payment to be processed
	time.Sleep(2 * time.Second)

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Set the partner reference number from the created payment
	jsonDict["originalPartnerReferenceNo"] = partnerReferenceNo
	jsonDict["partnerRefundNo"] = partnerReferenceNo

	// Marshal to JSON and unmarshal to widget SDK struct
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	var request widget.RefundOrderRequest
	err = json.Unmarshal(jsonBytes, &request)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute the SDK API call and expect error response
	ctx := context.Background()
	_, httpResponse, err := helper.ApiClient.WidgetAPI.RefundOrder(ctx).RefundOrderRequest(request).Execute()
	if err != nil {
		// This is expected for error test cases
		variableDict := map[string]interface{}{
			"partnerReferenceNo": partnerReferenceNo,
		}

		// Assert the error response matches expected error pattern
		err = helper.AssertFailResponse(refundOrderJsonPath, refundOrderTitleCase, caseName, httpResponse, variableDict)
		if err != nil {
			t.Fatal(err)
		}
	} else {
		// If no error occurred, this is unexpected for error test cases
		defer httpResponse.Body.Close()
		t.Fatalf("Expected error for case %s but API call succeeded", caseName)
	}
}

func TestRefundFailMandatoryParameterInvalid(t *testing.T) {
	caseName := "RefundFailMandatoryParameterInvalid"

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Marshal to JSON and unmarshal to widget SDK struct
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	var request widget.RefundOrderRequest
	err = json.Unmarshal(jsonBytes, &request)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute API request with missing mandatory field to trigger validation error
	ctx := context.Background()
	endpoint := "https://api.sandbox.dana.id/v1.0/debit/refund.htm"
	resourcePath := "/v1.0/debit/refund.htm"

	// Set custom headers with empty timestamp to trigger invalid mandatory field error
	customHeaders := map[string]string{
		"X-TIMESTAMP": "",
	}

	variableDict := map[string]interface{}{
		"partnerReferenceNo": jsonDict["originalPartnerReferenceNo"],
	}

	err = helper.ExecuteAndAssertErrorResponse(
		t,
		ctx,
		&request,
		"POST",
		endpoint,
		resourcePath,
		refundOrderJsonPath,
		refundOrderTitleCase,
		caseName,
		customHeaders,
		variableDict,
	)
	if err != nil {
		t.Fatal(err)
	}
}
func TestRefundFailOrderNotExist(t *testing.T) {
	caseName := "RefundFailOrderNotExist"

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Use the original non-existent partner reference number from the JSON
	// This should trigger the "Invalid Bill" error

	// Marshal to JSON and unmarshal to widget SDK struct
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	var request widget.RefundOrderRequest
	err = json.Unmarshal(jsonBytes, &request)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute the SDK API call and expect error response
	ctx := context.Background()
	_, httpResponse, err := helper.ApiClient.WidgetAPI.RefundOrder(ctx).RefundOrderRequest(request).Execute()
	if err != nil {
		// This is expected for error test cases
		variableDict := map[string]interface{}{
			"partnerReferenceNo": jsonDict["originalPartnerReferenceNo"],
		}

		// Assert the error response matches expected error pattern
		err = helper.AssertFailResponse(refundOrderJsonPath, refundOrderTitleCase, caseName, httpResponse, variableDict)
		if err != nil {
			t.Fatal(err)
		}
	} else {
		// If no error occurred, this is unexpected for error test cases
		defer httpResponse.Body.Close()
		t.Fatalf("Expected error for case %s but API call succeeded", caseName)
	}
}
func TestRefundFailInvalidSignature(t *testing.T) {
	caseName := "RefundFailInvalidSignature"

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Marshal to JSON and unmarshal to widget SDK struct
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	var request widget.RefundOrderRequest
	err = json.Unmarshal(jsonBytes, &request)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute API request with invalid signature to trigger unauthorized error
	ctx := context.Background()
	endpoint := "https://api.sandbox.dana.id/v1.0/debit/refund.htm"
	resourcePath := "/v1.0/debit/refund.htm"

	// Set custom headers with invalid signature to trigger unauthorized error
	customHeaders := map[string]string{
		"X-SIGNATURE": "invalid_signature",
	}

	variableDict := map[string]interface{}{
		"partnerReferenceNo": jsonDict["originalPartnerReferenceNo"],
	}

	err = helper.ExecuteAndAssertErrorResponse(
		t,
		ctx,
		&request,
		"POST",
		endpoint,
		resourcePath,
		refundOrderJsonPath,
		refundOrderTitleCase,
		caseName,
		customHeaders,
		variableDict,
	)
	if err != nil {
		t.Fatal(err)
	}
}

func TestRefundFailTimeout(t *testing.T) {
	caseName := "RefundFailTimeout"

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Marshal to JSON and unmarshal to widget SDK struct
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	var request widget.RefundOrderRequest
	err = json.Unmarshal(jsonBytes, &request)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute the SDK API call and expect error response
	ctx := context.Background()
	_, httpResponse, err := helper.ApiClient.WidgetAPI.RefundOrder(ctx).RefundOrderRequest(request).Execute()
	if err != nil {
		// This is expected for error test cases
		variableDict := map[string]interface{}{
			"partnerReferenceNo": jsonDict["originalPartnerReferenceNo"],
		}

		// Assert the error response matches expected error pattern
		err = helper.AssertFailResponse(refundOrderJsonPath, refundOrderTitleCase, caseName, httpResponse, variableDict)
		if err != nil {
			t.Fatal(err)
		}
	} else {
		// If no error occurred, this is unexpected for error test cases
		defer httpResponse.Body.Close()
		t.Fatalf("Expected error for case %s but API call succeeded", caseName)
	}
}
func TestRefundFailIdempotent(t *testing.T) {
	t.Skip("Skip: Idempotent test may require special handling or actual payment scenario")
	caseName := "RefundFailIdempotent"

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Marshal to JSON and unmarshal to widget SDK struct
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	var request widget.RefundOrderRequest
	err = json.Unmarshal(jsonBytes, &request)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute the SDK API call and expect success response (idempotent case)
	ctx := context.Background()
	apiResponse, httpResponse, err := helper.ApiClient.WidgetAPI.RefundOrder(ctx).RefundOrderRequest(request).Execute()
	if err != nil {
		t.Fatalf("API request failed: %v", err)
	}
	defer httpResponse.Body.Close()

	// Convert the response to JSON for assertion
	responseJSON, err := apiResponse.MarshalJSON()
	if err != nil {
		t.Fatalf("Failed to convert response to JSON: %v", err)
	}

	// Assert the success response
	variableDict := map[string]interface{}{
		"partnerReferenceNo": jsonDict["originalPartnerReferenceNo"],
	}

	err = helper.AssertResponse(
		refundOrderJsonPath,
		refundOrderTitleCase,
		caseName,
		string(responseJSON),
		variableDict,
	)
	if err != nil {
		t.Fatal(err)
	}
}
func TestRefundFailMerchantStatusAbnormal(t *testing.T) {
	caseName := "RefundFailMerchantStatusAbnormal"

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(refundOrderJsonPath, refundOrderTitleCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Marshal to JSON and unmarshal to widget SDK struct
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	var request widget.RefundOrderRequest
	err = json.Unmarshal(jsonBytes, &request)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute the SDK API call and expect error response
	ctx := context.Background()
	_, httpResponse, err := helper.ApiClient.WidgetAPI.RefundOrder(ctx).RefundOrderRequest(request).Execute()
	if err != nil {
		// This is expected for error test cases
		variableDict := map[string]interface{}{
			"partnerReferenceNo": jsonDict["originalPartnerReferenceNo"],
		}

		// Assert the error response matches expected error pattern
		err = helper.AssertFailResponse(refundOrderJsonPath, refundOrderTitleCase, caseName, httpResponse, variableDict)
		if err != nil {
			t.Fatal(err)
		}
	} else {
		// If no error occurred, this is unexpected for error test cases
		defer httpResponse.Body.Close()
		t.Fatalf("Expected error for case %s but API call succeeded", caseName)
	}
}

File: test/node/widget/refund_order_test.ts
import Dana, { ResponseError } from 'dana-node';
import { v4 as uuidv4 } from 'uuid';
import * as path from 'path';
import * as dotenv from 'dotenv';
import { fail } from 'assert';
import { getRequest } from '../helper/util';
import { assertResponse, assertFailResponse } from '../helper/assertion';
import { RefundOrderRequest } from 'dana-node/widget/v1';
import { executeManualApiRequest } from '../helper/apiHelpers';
import {
    createTestWidgetPaymentInit,
    createTestWidgetPaymentPaid,
} from './payment_widget_util';

dotenv.config();

const titleCase = 'RefundOrder';
const jsonPathFile = path.resolve(__dirname, '../../../resource/request/components/Widget.json');
const baseUrl = 'https://api.sandbox.dana.id/';
const apiPath = '/v1.0/debit/refund.htm';
const merchantId = process.env.MERCHANT_ID || '';

const dana = new Dana({
    partnerId: process.env.X_PARTNER_ID || '',
    privateKey: process.env.PRIVATE_KEY || '',
    origin: process.env.ORIGIN || '',
    env: process.env.ENV || 'sandbox',
});

function generateReferenceNo(): string {
    return uuidv4();
}

describe('RefundOrder Tests', () => {
    test('should successfully refund order (valid scenario)', async () => {
        const caseName = 'RefundOrderValidScenario';
        const partnerReferenceNo = await createTestWidgetPaymentPaid();
        await new Promise((resolve) => setTimeout(resolve, 2000));

        const requestData = getRequest<RefundOrderRequest>(jsonPathFile, titleCase, caseName);
        requestData.originalPartnerReferenceNo = partnerReferenceNo;
        requestData.partnerRefundNo = partnerReferenceNo;
        requestData.merchantId = merchantId;

        const response = await dana.widgetApi.refundOrder(requestData);
        await assertResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response), {
            partnerReferenceNo,
        });
    });

    test.skip('should successfully refund order (in process)', async () => {
        const caseName = 'RefundInProcess';
        const requestData: RefundOrderRequest = getRequest(jsonPathFile, titleCase, caseName);
        requestData.partnerRefundNo = generateReferenceNo();

        try {
            const response = await dana.widgetApi.refundOrder(requestData);
            await assertResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response));
        } catch (e: any) {
            if (e instanceof ResponseError) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            } else {
                fail('RefundOrder test failed: ' + (e.message || e));
            }
        }
    });

    test('should fail with duplicate request', async () => {
        const caseName = 'RefundFailDuplicateRequest';
        const partnerReferenceNo = await createTestWidgetPaymentPaid();
        await new Promise((resolve) => setTimeout(resolve, 2000));

        const requestData = getRequest<RefundOrderRequest>(jsonPathFile, titleCase, caseName);
        requestData.originalPartnerReferenceNo = partnerReferenceNo;
        requestData.partnerRefundNo = partnerReferenceNo;
        requestData.merchantId = merchantId;

        await dana.widgetApi.refundOrder(requestData);
        await new Promise((resolve) => setTimeout(resolve, 2000));

        requestData.refundAmount = { currency: 'IDR', value: '2.00' };

        try {
            await dana.widgetApi.refundOrder(requestData);
            fail('Expected duplicate refund error but API call succeeded');
        } catch (e: any) {
            if (e instanceof ResponseError) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse), {
                    partnerReferenceNo,
                });
            } else {
                fail('RefundOrder test failed: ' + (e.message || e));
            }
        }
    });

    test('should fail with order not paid', async () => {
        const caseName = 'RefundFailOrderNotPaid';
        const partnerReferenceNo = await createTestWidgetPaymentInit();
        await new Promise((resolve) => setTimeout(resolve, 2000));

        const requestData = getRequest<RefundOrderRequest>(jsonPathFile, titleCase, caseName);
        requestData.originalPartnerReferenceNo = partnerReferenceNo;
        requestData.partnerRefundNo = partnerReferenceNo;
        requestData.merchantId = merchantId;

        try {
            await dana.widgetApi.refundOrder(requestData);
            fail('Expected error but the API call succeeded');
        } catch (e: any) {
            if (e instanceof ResponseError) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse), {
                    originalPartnerReferenceNo: partnerReferenceNo,
                });
            } else {
                fail('RefundOrder test failed: ' + (e.message || e));
            }
        }
    });

    test('should fail with mandatory parameter invalid', async () => {
        const caseName = 'RefundFailMandatoryParameterInvalid';
        const requestData: RefundOrderRequest = getRequest(jsonPathFile, titleCase, caseName);
        const customerHeaders: Record<string, string> = {
            'X-TIMESTAMP': '',
        };

        try {
            await executeManualApiRequest(
                caseName,
                'POST',
                baseUrl + apiPath,
                apiPath,
                requestData,
                customerHeaders,
            );
            fail('Expected an error but the API call succeeded');
        } catch (e: any) {
            if (e instanceof ResponseError) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            } else if (Number(e.status) === 400) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            } else {
                fail('RefundOrder test failed: ' + (e.message || e));
            }
        }
    });

    test('should fail with order not exist', async () => {
        const caseName = 'RefundFailOrderNotExist';
        const requestData: RefundOrderRequest = getRequest(jsonPathFile, titleCase, caseName);

        try {
            const response = await dana.widgetApi.refundOrder(requestData);
            await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response));
            fail('Expected an error but the API call succeeded');
        } catch (e: any) {
            if (e instanceof ResponseError) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            } else {
                fail('RefundOrder test failed: ' + (e.message || e));
            }
        }
    });

    test('should fail with invalid signature', async () => {
        const caseName = 'RefundFailInvalidSignature';
        const requestData: RefundOrderRequest = getRequest(jsonPathFile, titleCase, caseName);
        const customerHeaders: Record<string, string> = {
            'X-SIGNATURE': 'invalid_signature',
        };

        try {
            await executeManualApiRequest(
                caseName,
                'POST',
                baseUrl + apiPath,
                apiPath,
                requestData,
                customerHeaders,
            );
            fail('Expected an error but the API call succeeded');
        } catch (e: any) {
            if (e instanceof ResponseError) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            } else if (Number(e.status) === 401) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            } else {
                fail('RefundOrder test failed: ' + (e.message || e));
            }
        }
    });

    test('should fail with timeout', async () => {
        const caseName = 'RefundFailTimeout';
        const requestData: RefundOrderRequest = getRequest(jsonPathFile, titleCase, caseName);
        requestData.partnerRefundNo = generateReferenceNo();

        try {
            const response = await dana.widgetApi.refundOrder(requestData);
            await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response));
            fail('Expected an error but the API call succeeded');
        } catch (e: any) {
            if (e instanceof ResponseError) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            } else {
                fail('RefundOrder test failed: ' + (e.message || e));
            }
        }
    });

    test.skip('should fail with idempotent', async () => {
        const caseName = 'RefundFailIdempotent';
        const requestData: RefundOrderRequest = getRequest(jsonPathFile, titleCase, caseName);
        fail('RefundOrder idempotent test skipped (same as Go)');
    });

    test('should fail with merchant status abnormal', async () => {
        const caseName = 'RefundFailMerchantStatusAbnormal';
        const requestData: RefundOrderRequest = getRequest(jsonPathFile, titleCase, caseName);
        requestData.partnerRefundNo = generateReferenceNo();

        try {
            const response = await dana.widgetApi.refundOrder(requestData);
            await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(response));
            fail('Expected an error but the API call succeeded');
        } catch (e: any) {
            if (e instanceof ResponseError) {
                await assertFailResponse(jsonPathFile, titleCase, caseName, JSON.stringify(e.rawResponse));
            } else {
                fail('RefundOrder test failed: ' + (e.message || e));
            }
        }
    });
});
